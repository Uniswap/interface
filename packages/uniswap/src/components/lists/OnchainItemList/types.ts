import { OnchainItemListOption } from 'uniswap/src/components/lists/items/types'

export enum OnchainItemSectionName {
  SearchResults = 'searchResults',
  RecentSearches = 'recentSearches',

  YourTokens = 'yourTokens',
  TrendingTokens = 'trendingTokens',
  FavoriteTokens = 'favoriteTokens',
  SuggestedTokens = 'suggestedTokens',
  BridgingTokens = 'bridgingTokens',
  OtherChainsTokens = 'otherNetworksTokens',

  Tokens = 'tokens',
  Pools = 'pools',
  TrendingPools = 'trendingPools',
  Wallets = 'wallets',
  FavoriteWallets = 'favoriteWallets',
  NFTCollections = 'nftCollections',
  PopularNFTCollections = 'popularNFTCollections',
}

export type OnchainItemSection<T extends OnchainItemListOption> = {
  data: T[]
  sectionKey: OnchainItemSectionName
  name?: string
  rightElement?: JSX.Element
  endElement?: JSX.Element
}
