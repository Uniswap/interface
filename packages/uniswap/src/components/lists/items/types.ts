import { ProtocolVersion } from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'
import { UniverseChainId } from 'uniswap/src/features/chains/types'
import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'

/* Types of list item options */
export enum OnchainItemListOptionType {
  Token = 'Token',
  Pool = 'Pool',
  WalletByAddress = 'WalletByAddress',
  ENSAddress = 'ENSAddress',
  Unitag = 'Unitag',
  NFTCollection = 'NFTCollection',
}

export interface BaseOption {
  type: OnchainItemListOptionType
}

export interface TokenOption extends BaseOption {
  type: OnchainItemListOptionType.Token
  currencyInfo: CurrencyInfo
  quantity: number | null // float representation of balance, returned by data-api
  balanceUSD: Maybe<number>
  isUnsupported?: boolean
}

export interface PoolOption extends BaseOption {
  type: OnchainItemListOptionType.Pool
  poolId: string
  chainId: UniverseChainId
  token0CurrencyInfo: CurrencyInfo
  token1CurrencyInfo: CurrencyInfo
  protocolVersion: ProtocolVersion
  hookAddress?: string
  feeTier: number
}

export type WalletOption = WalletByAddressOption | ENSAddressOption | UnitagOption

export interface WalletByAddressOption extends BaseOption {
  type: OnchainItemListOptionType.WalletByAddress
  address: Address
}
export interface ENSAddressOption extends BaseOption {
  type: OnchainItemListOptionType.ENSAddress
  address: Address
  isRawName?: boolean
  ensName: string
  primaryENSName?: string
}

export interface UnitagOption extends BaseOption {
  type: OnchainItemListOptionType.Unitag
  address: Address
  unitag: string
}

export interface NFTCollectionOption extends BaseOption {
  type: OnchainItemListOptionType.NFTCollection
  chainId: UniverseChainId
  address: Address
  name: string
  imageUrl: string | null
  isVerified: boolean
}
// Union of item types for different list use cases
export type MobileExploreSearchModalOption = TokenOption | WalletOption | NFTCollectionOption
export type WebSearchModalOption = TokenOption | PoolOption
export type SearchModalOption = MobileExploreSearchModalOption | WebSearchModalOption

export type TokenSelectorOption = TokenOption | TokenOption[]

// All item types combined
export type OnchainItemListOption = TokenSelectorOption | SearchModalOption
