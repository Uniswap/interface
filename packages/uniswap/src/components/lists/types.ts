import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'

/* Types of list item options */
export type TokenOption = {
  currencyInfo: CurrencyInfo
  quantity: number | null // float representation of balance, returned by data-api
  balanceUSD: Maybe<number>
  isUnsupported?: boolean
}

// Union of item types for different list use cases
export type SearchModalItemTypes = TokenOption // TODO: will add PoolOption
export type TokenSelectorItemTypes = TokenOption | TokenOption[]

// All item types combined
export type ItemType = TokenSelectorItemTypes | SearchModalItemTypes
