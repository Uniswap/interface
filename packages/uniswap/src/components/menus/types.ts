// How the menu is opened from the trigger element.
export enum ContextMenuTriggerMode {
  Primary = 'primary', // Left click on web. Regular press on mobile.
  Secondary = 'secondary', // Right click on web. Long press on mobile.
}
