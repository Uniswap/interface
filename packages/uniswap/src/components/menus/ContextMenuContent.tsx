import { Fragment } from 'react'
import { DropdownMenuSheetItem, DropdownMenuSheetItemProps, Flex, Separator, isWeb } from 'ui/src'
import { MenuOptionItem } from 'uniswap/src/components/menus/ContextMenuV2'

type MenuContentProps = {
  items: MenuOptionItem[]
  handleCloseMenu?: DropdownMenuSheetItemProps['handleCloseMenu']
}

export function MenuContent({ items, handleCloseMenu }: MenuContentProps): JSX.Element {
  return (
    <Flex
      asChild
      flexDirection="column"
      gap="$spacing4"
      p="$spacing8"
      backgroundColor="$surface1"
      borderRadius="$rounded20"
      borderWidth="$spacing1"
      borderColor="$surface3"
      minWidth={200}
    >
      {/* eslint-disable-next-line react/forbid-elements */}
      <div
        // Prevent any right-click from bubbling up or showing default context menu
        onContextMenu={(e) => {
          e.preventDefault()
          e.stopPropagation()
        }}
      >
        {items.map(({ label, onPress, Icon, showDivider, disabled, iconColor, closeDelay }, index) => (
          <Fragment key={index}>
            {showDivider && <Separator my="$spacing6" />}
            <DropdownMenuSheetItem
              variant={isWeb ? 'small' : 'medium'}
              label={label}
              icon={Icon && <Icon size="$icon.16" color={iconColor ?? '$neutral2'} />}
              disabled={disabled}
              closeDelay={closeDelay}
              handleCloseMenu={handleCloseMenu}
              onPress={onPress}
            />
          </Fragment>
        ))}
      </div>
    </Flex>
  )
}
