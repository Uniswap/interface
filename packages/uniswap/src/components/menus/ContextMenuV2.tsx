import { PropsWithChildren } from 'react'
import { GeneratedIcon, IconProps, TextProps } from 'ui/src'
import { ContextMenuTriggerMode } from 'uniswap/src/components/menus/types'
import { PlatformSplitStubError } from 'utilities/src/errors'

export type MenuOptionItem = {
  label: string
  onPress: () => void
  Icon?: GeneratedIcon | ((props: IconProps) => JSX.Element)
  showDivider?: boolean
  disabled?: boolean
  iconColor?: IconProps['color']
  textColor?: TextProps['color']
  closeDelay?: number
}

/**
 * Props for the ContextMenu component
 * @property menuItems - the menu options
 * @property isPlacementAbove / isPlacementRight - where the menu should be aligned relative to the trigger. By default,the menu is aligned to the bottom left of the trigger.
 * @property offsetX / offsetY - the offset of the menu from the trigger. By default, these are set to 0.
 * @property onPressAny - called when any menu option is pressed. will not be called if the onPress prop of the action throws
 * @property triggerMode - How the menu is triggered - primary (left click/tap) or secondary (right click/long press)
 * @property disabled
 * @property isOpen
 * @property closeMenu
 * @property openMenu - required if child component does not control menu state. leave undefined if the child component trigger opens the menu
 */
export type ContextMenuProps = {
  menuItems: MenuOptionItem[]
  isPlacementAbove?: boolean
  isPlacementRight?: boolean
  offsetX?: number
  offsetY?: number
  onPressAny?: (e: { name: string; index: number; indexPath: number[] }) => void
  triggerMode: ContextMenuTriggerMode
  disabled?: boolean
  isOpen: boolean
  closeMenu: () => void
  openMenu?: () => void
}

export function ContextMenu(_: PropsWithChildren<ContextMenuProps>): JSX.Element {
  throw new PlatformSplitStubError('ContextMenu')
}
