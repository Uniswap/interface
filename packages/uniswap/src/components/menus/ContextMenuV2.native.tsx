import isEqual from 'lodash/isEqual'
import { Fragment, PropsWithChildren, useCallback, useLayoutEffect, useMemo, useRef, useState } from 'react'
import { View } from 'react-native'
import { AnimatePresence, Flex, Portal, Separator, TouchableArea, useWindowDimensions } from 'ui/src'
import { DropdownMenuSheetItem } from 'ui/src/components/dropdownMenuSheet/DropdownMenuSheetItem'
import { zIndexes } from 'ui/src/theme'
import { ContextMenuProps } from 'uniswap/src/components/menus/ContextMenuV2'
import { ContextMenuTriggerMode } from 'uniswap/src/components/menus/types'
import { logger } from 'utilities/src/logger/logger'

// used for positioning
const CONTEXT_MENU_WIDTH = 205
const MIN_MENU_PADDING = 15
const MENU_OPTION_HEIGHT = 40
const MENU_OPTION_GAP = 4
const MENU_PADDING = 8

// used for animation
const ANIMATION_START_POINT = 10
const ANIMATION_TIME = 200

/**
 * A controlled styled context menu component.
 * Accepts both a onPress prop for each action and a onPressAny prop that is called when any action is pressed.
 * TODO(WALL-3692): replace the native context menu once it covers all current use cases.
 * @param children the trigger element
 * @returns a fragment with a context menu and a trigger
 */
export function ContextMenu({
  children,
  menuItems,
  isPlacementAbove = false,
  isPlacementRight = false,
  offsetX = 0,
  offsetY = 0,
  onPressAny,
  triggerMode,
  disabled = false,
  isOpen,
  closeMenu,
  openMenu,
}: PropsWithChildren<ContextMenuProps>): JSX.Element {
  const { width: screenWidth, height: screenHeight } = useWindowDimensions()

  const [isAboveTrigger, setIsAboveTrigger] = useState(isPlacementAbove)

  const isLongPress = triggerMode === ContextMenuTriggerMode.Secondary
  const triggerRef = useRef<View>(null)

  // used to control the visibility of the menu to allow for position calculations to complete before rendering
  const [isMenuVisible, setIsMenuVisible] = useState(false)

  const handleMenuClose = useCallback(() => {
    // used to delay unmount of the menu until the animation is done
    closeMenu()
    setTimeout(() => {
      setIsMenuVisible(false)
    }, ANIMATION_TIME)
  }, [closeMenu, setIsMenuVisible])

  const [position, setPosition] = useState<{
    left: number | undefined
    top: number | undefined
  }>({ left: 0, top: 0 })

  const recalculateMenuPosition = useCallback((): void => {
    if (isOpen && triggerRef.current) {
      triggerRef.current.measure((_fx, _fy, triggerWidth, triggerHeight, triggerX, triggerY) => {
        const maxUsableWidth = screenWidth - MIN_MENU_PADDING

        const getLeft = (): number => {
          const left: number = isPlacementRight
            ? triggerX + triggerWidth + offsetX // align *left* edge of menu to *right* edge of trigger
            : triggerX + triggerWidth - offsetX - CONTEXT_MENU_WIDTH // align *right* edge of menu to *right* edge of trigger

          // if menu overflows too far off the screen, clamp to edge of screen +/- MIN_MENU_PADDING
          if (left > maxUsableWidth - CONTEXT_MENU_WIDTH) {
            return maxUsableWidth - CONTEXT_MENU_WIDTH
          } else if (left < MIN_MENU_PADDING) {
            return MIN_MENU_PADDING
          }
          return left
        }

        // lazy eval to avoid unnecessary calculations
        const estimatedMenuHeight =
          menuItems.length * MENU_OPTION_HEIGHT + MENU_OPTION_GAP * (menuItems.length - 1) + MENU_PADDING * 2

        const getTop = (): number => {
          const aboveTriggerY: number = triggerY - estimatedMenuHeight - offsetY
          const belowTriggerY: number = triggerY + triggerHeight + offsetY

          if (aboveTriggerY < MIN_MENU_PADDING) {
            // if the menu overflows too far up off the screen, display below trigger
            setIsAboveTrigger(false)
            return belowTriggerY
          } else if (belowTriggerY + estimatedMenuHeight > screenHeight - MIN_MENU_PADDING) {
            // if the menu overflows too far down off the screen, display above trigger
            setIsAboveTrigger(true)
            return aboveTriggerY
          } else if (isPlacementAbove) {
            return aboveTriggerY
          } else {
            return belowTriggerY
          }
        }
        const left = getLeft()
        const top = getTop()
        setPosition((prev) => {
          const updated = { ...prev }

          updated.left = left
          updated.top = top

          // prevent unnecessary re-renders if the position has not changed
          if (isEqual(prev, { left, top })) {
            return prev
          }

          return updated
        })

        setIsMenuVisible(true)
      })
    }
  }, [isOpen, screenWidth, menuItems.length, isPlacementRight, offsetX, offsetY, isPlacementAbove, screenHeight])

  useLayoutEffect(() => {
    recalculateMenuPosition()
  }, [recalculateMenuPosition])

  const menuSheetItems = useMemo(() => {
    return menuItems?.map(
      (
        { label, textColor, Icon, iconColor, disabled: itemDisabled, onPress: onPressAction, showDivider, closeDelay },
        index,
      ) => (
        <Fragment key={index}>
          {showDivider && <Separator my="$spacing6" />}
          <DropdownMenuSheetItem
            key={index}
            variant="medium"
            label={label}
            textColor={textColor}
            icon={Icon && <Icon size="$icon.24" color={iconColor} />}
            height={MENU_OPTION_HEIGHT}
            disabled={itemDisabled}
            closeDelay={(closeDelay ?? 0) + ANIMATION_TIME}
            handleCloseMenu={() => {
              closeMenu()
              setIsMenuVisible(false)
            }}
            onPress={() => {
              try {
                // run both actions; `onPressAny` will not run if `onPressAction` throws
                onPressAction?.()
                onPressAny?.({ name: label, index, indexPath: [index] })
              } catch (error) {
                logger.error(error, {
                  tags: { file: 'ContextMenuV2.tsx', function: 'createPressHandler' },
                })
              }
            }}
          />
        </Fragment>
      ),
    )
  }, [closeMenu, menuItems, onPressAny])

  // the idea is that we cover the whole screen with a transparent area that closes the menu when pressed
  // and we have a child on top of that that is the actual menu
  // since only one of them can be pressed at a time, we don't have to worry about the event being propagated
  return (
    <>
      <Portal
        display={isOpen || isMenuVisible ? 'flex' : 'none'}
        contain="none"
        position="unset"
        // pass events through if menu is fading out
        pointerEvents={!isOpen ? 'none' : 'auto'}
        onPress={(e) => {
          e.stopPropagation()
        }}
      >
        <Flex
          height="100%"
          width="100%"
          top={0}
          left={0}
          backgroundColor="transparent"
          zIndex={zIndexes.overlay}
          onPress={handleMenuClose}
        >
          <AnimatePresence>
            {isMenuVisible && (
              <Flex
                justifyContent="flex-start"
                alignItems="flex-start"
                backgroundColor="$transparent"
                top={position.top}
                left={position.left}
                position="absolute"
                animation="200ms"
                enterStyle={{
                  opacity: 0,
                  y: isAboveTrigger ? ANIMATION_START_POINT : -ANIMATION_START_POINT,
                }}
                exitStyle={{
                  opacity: 0,
                  y: isAboveTrigger ? ANIMATION_START_POINT : -ANIMATION_START_POINT,
                }}
              >
                <Flex
                  backgroundColor="$surface1"
                  p={MENU_PADDING}
                  borderRadius="$rounded20"
                  borderColor="$surface3"
                  borderWidth="$spacing1"
                  gap={MENU_OPTION_GAP}
                  alignItems="flex-start"
                  width={CONTEXT_MENU_WIDTH}
                  shadowRadius="$spacing4"
                  shadowColor="$shadowColor"
                >
                  {menuSheetItems}
                </Flex>
              </Flex>
            )}
          </AnimatePresence>
        </Flex>
      </Portal>

      <Flex>
        {openMenu ? (
          <TouchableArea
            disabled={disabled}
            onPress={isLongPress ? undefined : openMenu}
            onLongPress={isLongPress ? openMenu : undefined}
          >
            <Flex ref={triggerRef} onLayout={recalculateMenuPosition}>
              {children}
            </Flex>
          </TouchableArea>
        ) : (
          // if openMenu is undefined, {children} controls menu open/close state. Don't want to interfere with nested TouchableAreas
          <Flex ref={triggerRef} onLayout={recalculateMenuPosition}>
            {children}
          </Flex>
        )}
      </Flex>
    </>
  )
}
