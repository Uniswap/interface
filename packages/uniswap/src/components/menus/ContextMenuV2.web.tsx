import { Fragment, PropsWithChildren, useRef, useState } from 'react'
import { Popover } from 'ui/src'
import { MenuContent } from 'uniswap/src/components/menus/ContextMenuContent'
import { ContextMenuProps } from 'uniswap/src/components/menus/ContextMenuV2'
import { ContextMenuTriggerMode } from 'uniswap/src/components/menus/types'
import { isMobileWeb } from 'utilities/src/platform'
import { useOnClickOutside } from 'utilities/src/react/hooks'

export function ContextMenu({
  menuItems,
  isPlacementAbove = false,
  isPlacementRight = false,
  offsetX = 0,
  offsetY = 0,
  triggerMode,
  disabled = false,
  children,
  isOpen,
  closeMenu,
  openMenu,
}: PropsWithChildren<ContextMenuProps>): JSX.Element {
  const containerRef = useRef<HTMLDivElement>(null)
  const triggerContainerRef = useRef<HTMLDivElement>(null)
  const [menuPosition, setMenuPosition] = useState({ x: 0, y: 0 })

  const isLeftClick = triggerMode === ContextMenuTriggerMode.Primary

  useOnClickOutside(containerRef, () => {
    closeMenu()
  })

  const onContextMenu = (e: React.MouseEvent<HTMLDivElement>): void => {
    if (disabled) {
      return
    }

    e.preventDefault()
    e.stopPropagation()
    openMenu?.()

    // Capture raw click coords
    const { clientX, clientY } = e
    setMenuPosition({ x: clientX, y: clientY })
  }

  function getRelativeCoordinates(): { x: number; y: number } {
    if (isLeftClick || !triggerContainerRef.current) {
      return { x: 0, y: 0 }
    }

    const rect = triggerContainerRef.current.getBoundingClientRect()
    const relativeX = isPlacementRight ? menuPosition.x - rect.left : menuPosition.x - rect.right
    const relativeY = menuPosition.y - rect.top - rect.height

    return {
      x: relativeX,
      y: relativeY,
    }
  }

  const { x, y } = getRelativeCoordinates()

  if (disabled || isMobileWeb) {
    return <Fragment>{children}</Fragment>
  }

  return (
    <Popover
      allowFlip
      open={isOpen}
      strategy="absolute"
      placement={
        isPlacementAbove
          ? isPlacementRight
            ? 'top-start' // above & to the right
            : 'top-end' // above & to the left
          : isPlacementRight
            ? 'bottom-start' // below & to the right
            : 'bottom-end' // below & to the left
      }
      offset={{
        mainAxis: y + (isPlacementAbove ? -offsetY : offsetY),
        crossAxis: x + (isPlacementRight ? offsetX : -offsetX),
      }}
    >
      {/*
        We attach the context menu event handler conditionally:
        - If onLeftClick is true, we use onMouseDown to open the menu on left-click.
        - Otherwise, we use onContextMenu to open the menu on right-click.
        This ensures that left-click events are not blocked from propagating,
        keeping normal click behavior intact.
      */}
      <Popover.Trigger onMouseDown={isLeftClick ? onContextMenu : undefined}>
        {/* eslint-disable-next-line react/forbid-elements */}
        <div ref={triggerContainerRef} onContextMenu={isLeftClick ? undefined : onContextMenu}>
          {children}
        </div>
      </Popover.Trigger>
      <Popover.Content
        ref={containerRef}
        key={`${menuPosition.x}-${menuPosition.y}`} // This key ensures that the component re-renders when the menu position changes so we get a re-animation
        backgroundColor="transparent"
        animation="125ms"
        enterStyle={{
          opacity: 0,
          scale: 0.98,
          transform: [{ translateY: -4 }],
        }}
      >
        <MenuContent items={menuItems} handleCloseMenu={closeMenu} />
      </Popover.Content>
    </Popover>
  )
}
