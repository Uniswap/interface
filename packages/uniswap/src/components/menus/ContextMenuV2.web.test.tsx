import { act, fireEvent } from '@testing-library/react-native'
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { View } from 'react-native'
import { Flex } from 'ui/src'
import { MenuOptionItem } from 'uniswap/src/components/menus/ContextMenuV2'
import { ContextMenu } from 'uniswap/src/components/menus/ContextMenuV2.web'
import { ContextMenuTriggerMode } from 'uniswap/src/components/menus/types'
import { renderWithProviders } from 'uniswap/src/test/render'

describe('ContextMenu', () => {
  const mockMenuItems: MenuOptionItem[] = [
    {
      label: 'Option 1',
      onPress: jest.fn(),
    },
    {
      label: 'Option 2',
      onPress: jest.fn(),
      showDivider: true,
    },
  ]

  it('renders without error', () => {
    const tree = renderWithProviders(
      <ContextMenu
        menuItems={mockMenuItems}
        triggerMode={ContextMenuTriggerMode.Secondary}
        isOpen={true}
        closeMenu={jest.fn()}
        openMenu={jest.fn()}
      >
        <Flex>Trigger</Flex>
      </ContextMenu>,
    )

    expect(tree).toMatchSnapshot()
  })

  describe('opens the menu', () => {
    it('on right-click', () => {
      const { getByTestId, queryByText } = renderWithProviders(
        <ContextMenu
          menuItems={mockMenuItems}
          triggerMode={ContextMenuTriggerMode.Secondary}
          isOpen={true}
          closeMenu={jest.fn()}
        >
          <View testID="trigger">
            <Flex>Trigger</Flex>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent(trigger, 'contextMenu', {
          preventDefault: jest.fn(),
          stopPropagation: jest.fn(),
          clientX: 100,
          clientY: 150,
        })
      })

      expect(queryByText('Option 1')).toBeTruthy()
      expect(queryByText('Option 2')).toBeTruthy()
    })

    it('on left-click', () => {
      const { getByTestId, queryByText } = renderWithProviders(
        <ContextMenu
          menuItems={mockMenuItems}
          triggerMode={ContextMenuTriggerMode.Primary}
          isOpen={true}
          closeMenu={jest.fn()}
        >
          <View testID="trigger">
            <Flex>Trigger</Flex>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent(trigger, 'onMouseDown', {
          preventDefault: jest.fn(),
          stopPropagation: jest.fn(),
          clientX: 100,
          clientY: 150,
        })
      })
      expect(queryByText('Option 1')).toBeTruthy()
      expect(queryByText('Option 2')).toBeTruthy()
    })
  })

  describe('handles edge cases', () => {
    it('does not open the menu if no menuItems are provided', () => {
      const { getByTestId, queryByRole } = renderWithProviders(
        <ContextMenu menuItems={[]} triggerMode={ContextMenuTriggerMode.Secondary} isOpen={true} closeMenu={jest.fn()}>
          <View testID="trigger">
            <Flex>Trigger</Flex>
          </View>
        </ContextMenu>,
      )

      const trigger = getByTestId('trigger')

      act(() => {
        fireEvent.press(trigger)
      })

      const menu = queryByRole('menu')
      expect(menu).toBeFalsy()
    })

    it('does not crash if trigger element is not found', () => {
      const { queryByRole } = renderWithProviders(
        <ContextMenu
          menuItems={mockMenuItems}
          triggerMode={ContextMenuTriggerMode.Secondary}
          isOpen={true}
          closeMenu={jest.fn()}
        >
          {null}
        </ContextMenu>,
      )

      const menu = queryByRole('menu')
      expect(menu).toBeFalsy()
    })
  })
})
