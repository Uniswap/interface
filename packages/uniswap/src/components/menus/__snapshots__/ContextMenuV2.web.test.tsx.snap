// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu renders without error 1`] = `
[
  <View
    aria-expanded={true}
    cancelable={true}
    disabled={false}
    focusable={true}
    minPressDuration={0}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
  >
    <div
      onContextMenu={[Function]}
    >
      <View
        style={
          {
            "flexDirection": "column",
          }
        }
      >
        Trigger
      </View>
    </div>
  </View>,
  <View
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 5001,
      }
    }
  >
    <View
      cancelable={true}
      disabled={false}
      focusable={true}
      minPressDuration={0}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "bottom": 0,
          "flexDirection": "column",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      pointerEvents="auto"
    >
      <View
        style={
          {
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "top": 0,
            "transform": [
              {
                "translateY": 0,
              },
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          id=":r3:"
          jestAnimatedStyle={
            {
              "value": {
                "backgroundColor": "transparent",
                "borderBottomLeftRadius": 0,
                "borderBottomRightRadius": 0,
                "borderTopLeftRadius": 0,
                "borderTopRightRadius": 0,
                "opacity": 0,
                "transform": [
                  {
                    "scale": 0.98,
                  },
                  {
                    "translateY": -4,
                  },
                ],
              },
            }
          }
          style={
            {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderBottomLeftRadius": 0,
              "borderBottomRightRadius": 0,
              "borderTopLeftRadius": 0,
              "borderTopRightRadius": 0,
              "flexDirection": "column",
              "opacity": 0,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
              "transform": [
                {
                  "scale": 0.98,
                },
                {
                  "translateY": -4,
                },
              ],
            }
          }
        >
          <div
            onContextMenu={[Function]}
            style={
              {
                "backgroundColor": "#131313",
                "borderBottomColor": "rgba(255,255,255,0.12)",
                "borderBottomLeftRadius": 20,
                "borderBottomRightRadius": 20,
                "borderBottomWidth": 1,
                "borderLeftColor": "rgba(255,255,255,0.12)",
                "borderLeftWidth": 1,
                "borderRightColor": "rgba(255,255,255,0.12)",
                "borderRightWidth": 1,
                "borderStyle": "solid",
                "borderTopColor": "rgba(255,255,255,0.12)",
                "borderTopLeftRadius": 20,
                "borderTopRightRadius": 20,
                "borderTopWidth": 1,
                "flexDirection": "column",
                "gap": 4,
                "minWidth": 200,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <View
              cancelable={true}
              disabled={false}
              focusable={true}
              hitSlop={[Function]}
              minPressDuration={0}
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onPress={[Function]}
              onPressIn={[Function]}
              onPressOut={[Function]}
              selectable={false}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#131313",
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "gap": 8,
                  "justifyContent": "space-between",
                  "maxWidth": 250,
                  "opacity": 1,
                  "paddingBottom": 8,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 8,
                  "transform": [
                    {
                      "scale": 1,
                    },
                  ],
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexShrink": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  maxFontSizeMultiplier={1.2}
                  numberOfLines={1}
                  style={
                    {
                      "color": "#FFFFFF",
                      "flexShrink": 1,
                      "fontFamily": "Basel Grotesk",
                      "fontSize": 17,
                      "fontWeight": "500",
                      "lineHeight": 19.549999999999997,
                      "opacity": 1,
                    }
                  }
                  suppressHighlighting={true}
                >
                  Option 1
                </Text>
              </View>
              <View
                style={
                  {
                    "flexDirection": "column",
                    "flexShrink": 0,
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    {
                      "height": 20,
                      "minHeight": 20,
                      "minWidth": 20,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                {
                  "borderBottomColor": "rgba(255,255,255,0.12)",
                  "borderBottomWidth": 1,
                  "borderLeftColor": "rgba(255,255,255,0.12)",
                  "borderLeftWidth": 0,
                  "borderRightColor": "rgba(255,255,255,0.12)",
                  "borderRightWidth": 0,
                  "borderStyle": "solid",
                  "borderTopColor": "rgba(255,255,255,0.12)",
                  "borderTopWidth": 0,
                  "flex": 1,
                  "flexShrink": 0,
                  "height": 0,
                  "marginBottom": 6,
                  "marginTop": 6,
                  "maxHeight": 0,
                }
              }
            />
            <View
              cancelable={true}
              disabled={false}
              focusable={true}
              hitSlop={[Function]}
              minPressDuration={0}
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onPress={[Function]}
              onPressIn={[Function]}
              onPressOut={[Function]}
              selectable={false}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#131313",
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "flexDirection": "row",
                  "flexGrow": 1,
                  "gap": 8,
                  "justifyContent": "space-between",
                  "maxWidth": 250,
                  "opacity": 1,
                  "paddingBottom": 8,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 8,
                  "transform": [
                    {
                      "scale": 1,
                    },
                  ],
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexShrink": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  maxFontSizeMultiplier={1.2}
                  numberOfLines={1}
                  style={
                    {
                      "color": "#FFFFFF",
                      "flexShrink": 1,
                      "fontFamily": "Basel Grotesk",
                      "fontSize": 17,
                      "fontWeight": "500",
                      "lineHeight": 19.549999999999997,
                      "opacity": 1,
                    }
                  }
                  suppressHighlighting={true}
                >
                  Option 2
                </Text>
              </View>
              <View
                style={
                  {
                    "flexDirection": "column",
                    "flexShrink": 0,
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    {
                      "height": 20,
                      "minHeight": 20,
                      "minWidth": 20,
                      "width": 20,
                    }
                  }
                />
              </View>
            </View>
          </div>
        </View>
      </View>
    </View>
  </View>,
]
`;
